### ISSUE TITLE
TF's statically linked OpenMP causes performance problems with other libraries

### ISSUE THREAD

**Issue by mergian** [tensorflow/tensorflow#56671]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Performance

### Source

binary

### Tensorflow Version

2.9.1

### Custom Code

No

### OS Platform and Distribution

Linux CentOS

### Python version

3.7

### Current Behaviour?

```shell
Hey TF team,

It seems that the PYPI version of TF is statically linked to its own OpenMP
implementation. This however can interfere when using multiple libraries that
rely on OpenMP.

The problem is, that the TF OpenMP does not free the threads immediately, but
spins for a certain amount of time, in case more work is coming in. When the
data is passed onto other libraries, these need to wait till TF put's it's
OpenMP threads to sleep. This causes the other library to wait, ruining the
performance of the entire pipeline. In our case the performance drop was 3-4x
compared to the pure runtime within TF and our library.

You can see that TF comes with it's own OpenMP implementation using
```OMP_DISPLAY_ENV=VERBOSE python3 ``` and the code posted below. It doesn't
matter if you use ```torch``` or any other Python library that uses the system
OpenMP. I removed unnecessary parts from the output.

This might be related to #29968.

A solution to this could be, that TF link to a shared version of OpenMP, i.e.,
as PyTorch and other packages are already doing it.
```

### Standalone code to reproduce the issue

```shell
print(">> SHARED OPENMP")
import torch
print("<< SHARED OPENMP")

import tensorflow
print(">> TF BUNDLED OPENMP")
with tensorflow.device('cpu'):
	pass
print("<< TF BUNDLED OPENMP")
```

### Relevant log output

```shell
>> SHARED OPENMP
	OPENMP DISPLAY ENVIRONMENT BEGIN
	  _OPENMP = '201511'
	  OMP_DYNAMIC = 'FALSE'
	  OMP_NESTED = 'FALSE'
	  OMP_NUM_THREADS = '24'
	  OMP_SCHEDULE = 'DYNAMIC'
	  OMP_PROC_BIND = 'TRUE'
	  OMP_PLACES = '{0},{1},{2},{3},{4},{5},{6},{7},{8},{9},{10},{11},{12},{13},{14},{15},{16},{17},{18},{19},{20},{21},{22},{23}'
	  OMP_STACKSIZE = '140265815980336'
	  OMP_WAIT_POLICY = 'PASSIVE'
	  OMP_THREAD_LIMIT = '4294967295'
	  OMP_MAX_ACTIVE_LEVELS = '2147483647'
	  OMP_CANCELLATION = 'FALSE'
	  OMP_DEFAULT_DEVICE = '0'
	  OMP_MAX_TASK_PRIORITY = '0'
	  GOMP_CPU_AFFINITY = ''
	  GOMP_STACKSIZE = '140265815980336'
	  GOMP_SPINCOUNT = '300000'
	OPENMP DISPLAY ENVIRONMENT END
<< SHARED OPENMP

...

>> TF BUNDLED OPENMP
	2022-07-04 09:47:08.193122: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA
	To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.

	OPENMP DISPLAY ENVIRONMENT BEGIN
	_OPENMP = 201511
	OMP_NUM_THREADS = 8
	OMP_STACKSIZE = 4194304
	OMP_SCHEDULE = static
	OMP_THREAD_LIMIT = 8
	OMP_MAX_ACTIVE_LEVELS = 1
	OMP_NESTED = false
	OMP_DYNAMIC = false
	OMP_WAIT_POLICY = active
	OMP_TOOL = disabled
	NOMP_MAX_QUEUED_TASKS = 1024
	NOMP_TASK_QUEUE_TYPE = HYBRID
	NOMP_TASK_QUEUE_ORDER = FIFO
	NOMP_TASK_REUSE_MEM = false
	OPENMP DISPLAY ENVIRONMENT END
<< TF BUNDLED OPENMP
```
```
</details>

**Comment by gowthamkpr** [comment]
Linking issue #29968 as this might be related.

**Comment by agramesh1** [comment]
@mergian are you using  the Intel version of the pypi  i.e pip install intel-tensorflow?
The official pypi Tensoflow doesn't use openMP  only the Intel build of Tensorflow uses openMP.  The Intel version includes its own version of LLVM's openMP as a shared library.

**Comment by mergian** [comment]
@agramesh1 no I'm using the official PYPI TensorFlow. I just created a blank VENV and only installed TF. It seems the second OpenMP output comes from a device support library we are using.

BUT I was able to reproduce the performance problems also with just a blank TF when running our pipeline on CPU, so I assume that the TF threading implementation itself might interfere with other libraries using OpenMP.
